### PR TITLE
refactor(logging): structlog where missing, shorter/cleaner messages

### PR DESCRIPTION
**Pull request by jansegre**
TODO:

- [x] change base to `dev` after `refactor/speed-initialization` is merged

This is what it looks like with this PR:

[asciicast](https://asciinema.org/a/HlJwgYFzbStVRINIj0CWHFHtj?speed=30&theme=tango)

And this is what it looked like before (thumbnail isn't great):

[asciicast](https://asciinema.org/a/kURqhQBfLuGoZPf4PWJ76uN80?speed=30&theme=tango)

### CODE REVIEW

File: hathor/p2p/node_sync.py [hunk 1]
```diff
@@ -355,23 +357,20 @@ def sync_from_timestamp(self, next_timestamp: int) -> Iterator[Deferred]:
 
         :param next_timestamp: Timestamp to start the sync
         """
-        self.log.debug('sync-{p} Sync starting at {next_timestamp}', p=self.short_peer_id,
-                       next_timestamp=next_timestamp)
+        self.log.debug('sync start', ts=next_timestamp)
```
Comment (obiyankenobi) [inline hathor/p2p/node_sync.py:360]: Why did you remove the peer id from all the logs in this file?

File: hathor/stratum/stratum.py [hunk 2]
```diff
@@ -523,8 +523,10 @@ def handle_submit(self, params: Dict, msgid: Optional[str]) -> None:
             tx.verify_pow()
         except PowError:
             # Transaction pow was not enough, but the share was succesfully submited
-            self.log.info('Insufficient weight, keep searching', tx=tx)
+            self.log.info('high hash, keep mining', tx=tx)
```
Comment (obiyankenobi) [inline hathor/stratum/stratum.py:526]: The original message seems better to me ("insufficient weight" x "high hash").

Comment (jansegre) [reply hathor/stratum/stratum.py:526]: "high hash" is the expression that Bitcoin uses, and one problem with "insufficient weight" is that it was often mistaken for an error, that happened a lot at the beginning when new people were mining with their own fullnodes

File: hathor/p2p/node_sync.py (hunk 1 above)
Comment (jansegre) [reply hathor/p2p/node_sync.py:360]: I moved it to `self.log = logger.new(peer_id_short=self.short_peer_id)`, with that every call to `self.log.*` will include `peer_id_short`.

Comment (obiyankenobi) [reply hathor/p2p/node_sync.py:360]: Perfect, thanks!

File: hathor/stratum/stratum.py (hunk 2 above)
Comment (obiyankenobi) [reply hathor/stratum/stratum.py:526]: Got it

File: hathor/manager.py [hunk 3]
```diff
@@ -290,37 +288,29 @@ def _initialize_components(self) -> None:
...
 
         # self.stop_profiler(save_to='profiles/initializing.prof')
         self.state = self.NodeState.READY
-        self.log.info(
-            'Node successfully initialized (total={total}, avg={avg:.2f} tx/s in {dt} seconds).',
-            total=cnt,
-            avg=cnt / (t2 - t0),
-            dt=t2 - t0,
-        )
+        self.log.info('ready', tx_count=cnt, tx_rate=cnt / (t2 - t0), total_dt=t2 - t0)
```
Comment (pedroferreira1) [inline hathor/manager.py:313]: I am still not sure why change 'Node successfully initialized' to 'ready'. Right now we have some things that could be ready, e.g. prometheus, metrics, stratum. I like being verbose in this case

File: tests/cli/test_multisig_address.py [hunk 4]
```diff
@@ -22,16 +22,16 @@ def test_generate_address(self):
         # Last element is always empty string
         output.pop()
 
-        self.assertEqual(len(output), pubkey_count * 10 + 11)
+        self.assertEqual(len(output), pubkey_count * 11 + 11)
```
Comment (pedroferreira1) [inline tests/cli/test_multisig_address.py:25]: Why this has changed?

Comment (jansegre) [reply tests/cli/test_multisig_address.py:25]: This test is kind of weird because it depends on the specific number of lines from the output. Because there is a new log output, the number of lines changed.

File: hathor/manager.py (hunk 3 above)
Comment (jansegre) [reply hathor/manager.py:313]: It's scoped by `hathor.manager`, so in my head this was fine.

File: tests/cli/test_multisig_address.py (hunk 4 above)
Comment (pedroferreira1) [reply tests/cli/test_multisig_address.py:25]: Ah, now I understand.

File: hathor/manager.py (hunk 3 above)
Comment (pedroferreira1) [reply hathor/manager.py:313]: Ah, right. I forgot about the scope, then it's fine.

As you said before, I think an image of the new logs would be good.

File: hathor/merged_mining/coordinator.py [hunk 5]
```diff
@@ -719,12 +719,11 @@ async def submit_to_bitcoin(self, job: SingleMinerJob, work: SingleMinerWork) ->
...
                 f.close()
-                self.log.warn('Work rejected from Bitcoin Fullnode. Relevant job+work dumped.',
-                              code=res, filename=f.name)
+                self.log.warn('block rejected from Bitcoin', code=res, dump_file=f.name)
             else:
-                self.log.warn('Work rejected from Bitcoin Fullnode.', code=res)
+                self.log.warn('block rejected from Bitcoin', code=res)
         else:
-            self.log.info('Bitcoin RPC submit response', res=res)
+            self.log.info('other result from Bitcoin', res=res)
```
Comment (msbrogli) [inline hathor/merged_mining/coordinator.py:726]: Maybe "unexpected result from Bitcoin"?

**Comment by jansegre** [comment]
If I comment `self.manager.start()` on `hathor.cli.run_log` and try to GET `/v1a/status` this is what the logs looks like with this PR:

```
2020-07-31 16:39:02 [info     ] [hathor.cli.run_node] hathor-core v0.36.1            genesis=3fdff62 my_peer_id=7181d232a16ba95001de22d3151b027ed31677c8cd6dc7eb88d538b250186bbf version=0.36.1
2020-07-31 16:39:02 [info     ] [hathor.cli.run_node] with storage                   path=/Users/jan/Projects/hathor-core/datamain storage_class=TransactionRocksDBStorage
2020-07-31 16:39:02 [info     ] [hathor.cli.run_node] with cache                     capacity=100000 interval=5
2020-07-31 16:39:07 [critical ] [twisted]                                
Traceback (most recent call last):
  File "/Users/jan/.local/share/virtualenvs/hathor-core-ygIHHmG7/lib/python3.6/site-packages/twisted/web/server.py", line 199, in process
    self.render(resrc)
  File "/Users/jan/.local/share/virtualenvs/hathor-core-ygIHHmG7/lib/python3.6/site-packages/twisted/web/server.py", line 259, in render
    body = resrc.render(self)
  File "/Users/jan/.local/share/virtualenvs/hathor-core-ygIHHmG7/lib/python3.6/site-packages/twisted/web/resource.py", line 250, in render
    return m(request)
  File "/Users/jan/Projects/hathor-core/hathor/p2p/resources/status.py", line 74, in render_GET
    'state': self.manager.state.value,
AttributeError: 'NoneType' object has no attribute 'value'
```

And then before it looked like this:

```
2020-07-31 16:45:06 [info     ] [twisted] Log opened.
2020-07-31 16:45:07 [info     ] [twisted] Hathor v0.36.1 (genesis 3fdff62)
2020-07-31 16:45:07 [info     ] [twisted] My peer id is 7181d232a16ba95001de22d3151b027ed31677c8cd6dc7eb88d538b250186bbf
2020-07-31 16:45:07 [info     ] [twisted] Using Wallet at /Users/jan/Projects/hathor-core/datamain
2020-07-31 16:45:07 [info     ] [twisted] Using TransactionRocksDBStorage at /Users/jan/Projects/hathor-core/datamain
2020-07-31 16:45:07 [info     ] [twisted] Using TransactionCacheStorage, capacity 100000, interval 5s
2020-07-31 16:45:07 [info     ] [twisted] Site starting on 9080
2020-07-31 16:45:07 [info     ] [twisted] Starting factory <twisted.web.server.Site object at 0x10fba7080>
2020-07-31 16:45:10 [critical ] [twisted]                                exc_info=(<class 'AttributeError'>, AttributeError("'NoneType' object has no attribute 'value'",), <traceback object at 0x10fba14c8>)
2020-07-31 16:45:10 [info     ] [twisted] "127.0.0.1" - - [31/Jul/2020:16:45:10 +0000] "GET /v1a/status HTTP/1.1" 500 4771 "-" "curl/7.64.1"
```

File: hathor/merged_mining/coordinator.py (hunk 5 above)
Comment (jansegre) [reply hathor/merged_mining/coordinator.py:726]: I'm not sure if it's unexpected. If it's not `None` (success) or `'high-hash'` they could all be called `unexpected` because it would signal an implementation error, actually even `'high-hash'` would be unexpected because of the `if block_hash.to_u256() > block_target.to_u256()` condition above. Maybe I could refactor these since I understand these errors better now, and because we'll validate the template before sending the work to the miner (which is on another branch).

Comment (msbrogli) [reply hathor/merged_mining/coordinator.py:726]: That would be great! You could also include a list of the known responses from the Bitcoin full node.

File: hathor/cli/util.py [hunk 6]
```diff
@@ -139,11 +156,15 @@ def __call__(self, _, __, event_dict):
...
+                    'handlers': ['default'],
+                    'level': 'INFO' if debug else 'WARN',
+                    'propagate': False,
+                },
                 '': {
-                    # 'handlers': ['default', 'file'],
                     'handlers': ['default'],
                     'level': 'DEBUG' if debug else 'INFO',
-                    'propagate': True,
+                    # 'propagate': True,
```
Comment (msbrogli) [inline hathor/cli/util.py:167]: Should we remove this line? Or add a comment explaining when it should be uncommented?

File: hathor/cli/util.py [hunk 7]
```diff
@@ -164,17 +185,37 @@ def kwargs_formatter(_, __, event_dict):
...
+            failure = event.get('log_failure')
+            if failure is not None:
+                kwargs['exc_info'] = (failure.type, failure.value, failure.getTracebackObject())
+            twisted_logger.log(level, msg, **kwargs)
+        except Exception as e:
+            print('error when logging event', e)
+
+    # start logging to std logger so structlog can catch it
+    twisted.python.log.startLoggingWithObserver(twisted_structlog_observer, setStdout=capture_stdout)
+    # twisted.python.log.startLoggingWithObserver(twisted_structlog_observer, setStdout=False)
```
Comment (msbrogli) [inline hathor/cli/util.py:215]: Remove this line?

File: hathor/cli/util.py [hunk 8]
```diff
@@ -164,17 +185,37 @@ def kwargs_formatter(_, __, event_dict):
...
+            if not msg and 'log_format' in event:
+                msg = event['log_format'].format(**event)
+            if not msg and 'format' in event:
+                msg = event['format'] % event
+            failure = event.get('log_failure')
+            if failure is not None:
+                kwargs['exc_info'] = (failure.type, failure.value, failure.getTracebackObject())
+            twisted_logger.log(level, msg, **kwargs)
+        except Exception as e:
+            print('error when logging event', e)
```
Comment (msbrogli) [inline hathor/cli/util.py:211]: Should we exit the process?

File: hathor/cli/util.py (hunk 6 above)
Comment (jansegre) [reply hathor/cli/util.py:167]: I'll link to python's logging docs, which explain what all theses settings mean and remove that line.

File: hathor/cli/util.py (hunk 7 above)
Comment (jansegre) [reply hathor/cli/util.py:215]: Yeah, I'll remove this one too.

File: hathor/cli/util.py (hunk 8 above)
Comment (jansegre) [reply hathor/cli/util.py:211]: I'd rather not. I have on idea what could cause this. And if it turn out to be "frequent" we're going to have a bad time managing our nodes that don't restart automatically and even if they did it would take a good while to load/sync and be up and running again.

Let's first see how it goes.

File: hathor/merged_mining/coordinator.py (hunk 5 above)
Comment (jansegre) [reply hathor/merged_mining/coordinator.py:726]: OK, I'll do this on `feat/merged-mining-better-api` where I'm already touching this part. Also, although it might be helpful to list known responses and symptoms, it's probably not that useful. It should always return `None`, otherwise there is an implementation error. At least that's what I get from all the errors I've encountered. But I'll add a brief list anyway so I can remember it better in case there is a regression.

File: hathor/cli/util.py (hunk 7 above)
Comment (jansegre) [reply hathor/cli/util.py]: Done.

File: hathor/cli/util.py (hunk 6 above)
Comment (jansegre) [reply hathor/cli/util.py]: Done. What do you think?

File: hathor/merged_mining/coordinator.py (hunk 5 above)
Comment (jansegre) [reply hathor/merged_mining/coordinator.py:726]: Here: https://github.com/HathorNetwork/hathor-core/pull/114/files#diff-eb36e58f1ad6ea122750ed239d9cd974L709-L729
